src/nupic/types/Sdr.hpp

```diff
@@ -213,10 +213,16 @@ class SparseDistributedRepresentation : public Serializable
      *
      * @param dimensions A list of dimension sizes, defining the shape of the
      * SDR.  The product of the dimensions must be greater than zero.
+     *
+     * @param (optional) data: default (sparse) data, same as calling sdr.setSparse(data);
+     * Default value is empty{}.
+     *
      */
-    SparseDistributedRepresentation( const std::vector<UInt> dimensions );
+    SparseDistributedRepresentation( const std::vector<UInt> dimensions,
+		                     const SDR_sparse_t data);
```

Comment:
I don't like this change.

* Initializing the SDR with data is not a common use case.  Unit tests are probably the only real application for this constructor.
   - We could make the unit tests easier to write by adding `operator==` overload to assign data from a sparse format vector.

* This change also makes it harder to use the SDR because now you need to remember to give it an empty vector.  `SDR( dimensions )` becomes `SDR( dimensions, {} )` and for most use cases the second vector is going to be empty.

* A simple dimensions check will detect the error, and the code already has dimension checks everywhere.  I don't think this is really a problem.

Reply:
> We could make the unit tests easier to write by adding operator== overload to assign data from a sparse format vector. 

this is a good idea :+1: 

> This change also makes it harder to use the SDR because now you need to remember to give it an empty vector. SDR( dimensions ) becomes SDR( dimensions, {} ) and for most use cases the second vector is going to be empty.

I'm open to closing this PR. 
But in my opinion having to type extra `{}` in SDR's constructor (rarely, since all the code is converted in this PR) isn't such a problem compared to the annoying bugs this still triggers from time to time. 

> A simple dimensions check will detect the error, and the code already has dimension checks everywhere. I don't think this is really a problem.
 
this bug bites when you refactor code, new fn expects SDR, something still gives vector, and this silently compiles.

Reply:
I'm going to close this PR.  The change is invasive and it solves a minor issue.  There are nicer ways to solve this problem, such as: better documentation, and more error checking.